[14.0][IMP] shopfloor: location content transfer find_work multiple user

mt-software-de opened the pull request:
Before:
Also pickings were found which are already assigned to a user This leaded to two problems:
- if the qty done was already set to the max qty by an other user you always got the error no work
- or you got the same location and the same move lines as the user before

Now:
You only get pickings whitout a assigned user

mt-software-de commented:
@jbaudoux this should solve the problem that the users getting always the same location until the previous picking isn't done

jbaudoux reviewed (approved):
Thanks for this fix.
To make it safe, there should also be a lock at the beginning of `_find_location_to_work_from` like here https://github.com/OCA/wms/blob/14.0/shopfloor_batch_automatic_creation/actions/picking_batch_auto_create.py#L59

mt-software-de commented:
> Thanks for this fix. To make it safe, there should also be a lock at the beginning of `_find_location_to_work_from` like here https://github.com/OCA/wms/blob/14.0/shopfloor_batch_automatic_creation/actions/picking_batch_auto_create.py#L59

done

simahawk commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -245,18 +247,27 @@ def _create_moves_from_location(self, location):
+        offset = -1
+        while not location:
+            offset += 1
+            next_picking = self.env["stock.picking"].search(
```
wouldn't be better to search them all just once and slice them w/ the offset?

simahawk commented on shopfloor_base/utils.py:
```diff
@@ -65,3 +68,17 @@ def _get_running_env():
 RUNNING_ENV = _get_running_env()
+
+
+def _lock(cr, name):
```
I would move this to an action. We can have `self._actions_for("lock").advisory(name)`.

simahawk replied:
we can have `for_update` too and replace `_lock_lines` in cluster_picking and location_content_transfer

mt-software-de replied:
Thanks for the input i wasn't aware of  the actions, this a better place to put it in.
I created the new action and replaced the calls

jbaudoux commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -273,6 +280,7 @@ def find_work(self):
         response = self._recover_started_picking()
         if response:
             return response
+        self._actions_for("lock").advisory("find_work")
```
Can the advisory lock name move to a variable or method? And I would give it a more unique name like location_content_transfer_find_work

jbaudoux commented on shopfloor_base/actions/lock.py:
```diff
@@ -0,0 +1,32 @@
+        self.env.cr.execute("SELECT pg_advisory_xact_lock(%s);", (int_lock,))
+        self.env.cr.fetchone()[0]
+
+    def for_update(self, table, ids, log_exceptions=False):
```
I would have given the recordset in argument instead of table & ids

simahawk commented on shopfloor_base/actions/lock.py:
```diff
@@ -0,0 +1,32 @@
+# Copyright 2022 Michael Tietz (MT Software) <mtietz@mt-software.de>
```
```suggestion
# Copyright 2020 Camptocamp
# Copyright 2022 Michael Tietz (MT Software) <mtietz@mt-software.de>
```

simahawk reviewed (commented):
LG really good, thanks.
One nitpicking thing: can you split the commit message pls?
Eg:
```
shopfloor_base: add actions for locking records

* advisory
* for_update
```

```
shopfloor: use lock actions
```

```
shopfloor: fix location content transfer find_work

Before:
Also pickings were found which are already assigned to a user
This leaded to two problems:
- if the qty done was already set to the max qty
  by an other user you always got the error no work
- or you got the same location and the same move lines as the user before

Now:
You only get pickings whithout a assigned user
```
The third one can contain the locking change for find_work, of course.

mt-software-de committed:
[IMP] shopfloor_base: add actions for locking records

* advisory
* for_update

mt-software-de committed:
[IMP] shoploor: use lock actions

mt-software-de committed:
[FIX] shopfloor: fix location content transfer find_work

Before:
Also pickings were found which are already assigned to a user
This leaded to two problems:
- if the qty done was already set to the max qty
  by an other user you always got the error no work
- or you got the same location and the same move lines as the user before

Now:
The find_work method gets locked, before finding a location.
Also it searches now for all assigned pickings, which are not marked as printed
and returns the location of the first open move_line

simahawk reviewed (approved):
Thanks for your improvements and your collaboration :+1:

simahawk commented:
/ocabot merge minor
